src/Repository/DeviceCodeRepository.php

```diff
@@ -0,0 +1,190 @@
+        $deviceCode = $this->deviceCodeManager->find($codeId);
+
+        if (null === $deviceCode) {
+            return true;
+        }
+
+        return $deviceCode->isRevoked();
+    }
+
+    private function buildDeviceCodeEntity(DeviceCodeInterface $deviceCode): DeviceCodeEntity
+    {
+        $deviceCodeEntity = new DeviceCodeEntity();
+        $deviceCodeEntity->setIdentifier($deviceCode->getIdentifier());
+        $deviceCodeEntity->setExpiryDateTime($deviceCode->getExpiry());
+        $client = $this->clientRepository->getClientEntity($deviceCode->getClient()->getIdentifier());
+        if ($client) {
+            $deviceCodeEntity->setClient($client);
+        }
+        if ($deviceCode->getUserIdentifier()) {
+            $deviceCodeEntity->setUserIdentifier($deviceCode->getUserIdentifier());
+        }
+        $deviceCodeEntity->setUserCode($deviceCode->getUserCode());
+        $deviceCodeEntity->setUserApproved($deviceCode->getUserApproved());
+        $deviceCodeEntity->setVerificationUriCompleteInAuthResponse($deviceCode->getIncludeVerificationUriComplete());
+        $deviceCodeEntity->setVerificationUri($deviceCode->getVerificationUri());
+        if ($deviceCode->getLastPolledAt()) {
+            $deviceCodeEntity->setLastPolledAt($deviceCode->getLastPolledAt());
+        }
+        $deviceCodeEntity->setInterval($deviceCode->getInterval());
+
+        foreach ($deviceCode->getScopes() as $scope) {
+            $deviceCodeEntity->addScope($this->scopeConverter->toLeague($scope));
+        }
+
+        return $deviceCodeEntity;
+    }
+
+    private function buildDeviceCodeModel(DeviceCodeEntityInterface $deviceCodeEntity): DeviceCodeModel
+    {
+        /** @var AbstractClient $client */
+        $client = $this->clientManager->find($deviceCodeEntity->getClient()->getIdentifier());
+
+        return new DeviceCodeModel(
+            $deviceCodeEntity->getIdentifier(),
+            $deviceCodeEntity->getExpiryDateTime(),
+            $client,
+            $deviceCodeEntity->getUserIdentifier(),
+            $this->scopeConverter->toDomainArray(array_values($deviceCodeEntity->getScopes())),
+            $deviceCodeEntity->getUserCode(),
+            $deviceCodeEntity->getUserApproved(),
+            $deviceCodeEntity->getVerificationUri(),
+            $deviceCodeEntity->getLastPolledAt(),
+            $deviceCodeEntity->getInterval()
+        );
+    }
+
+    private function updateDeviceCodeModel(
+        DeviceCodeInterface $deviceCode,
+        DeviceCodeEntityInterface $deviceCodeEntity,
+    ): DeviceCodeInterface {
+        if ($deviceCodeEntity->getLastPolledAt()) {
+            $deviceCode->setLastPolledAt($deviceCodeEntity->getLastPolledAt());
+        }
+
+        if ($deviceCodeEntity->getUserIdentifier()) {
+            $deviceCode->setUserIdentifier($deviceCodeEntity->getUserIdentifier());
+            $deviceCode->setUserApproved($deviceCodeEntity->getUserApproved());
+        }
+
+        return $deviceCode;
```

Comment:
As this method mutates the received `$deviceCode`, I think it should not return anything.

```suggestion
    private function updateDeviceCodeModel(
        DeviceCodeInterface $deviceCode,
        DeviceCodeEntityInterface $deviceCodeEntity,
    ): void {
        if ($deviceCodeEntity->getLastPolledAt()) {
            $deviceCode->setLastPolledAt($deviceCodeEntity->getLastPolledAt());
        }

        if ($deviceCodeEntity->getUserIdentifier()) {
            $deviceCode->setUserIdentifier($deviceCodeEntity->getUserIdentifier());
            $deviceCode->setUserApproved($deviceCodeEntity->getUserApproved());
        }
```